### ISSUE TITLE
Support TypeIs narrowing for exhaustiveness checking in pattern matching

### ISSUE THREAD

**Issue by sveinugu** [microsoft/pyright#10680]
**Is your feature request related to a problem? Please describe.**
Consider the following code example:

```python
from typing import Literal, TypeIs, Any, assert_never

def is_yes_or_no(choice: Any) -> TypeIs[Literal['yes', 'no']]:
    return choice in ['yes', 'no']

def check_choice(choice: Literal['yes', 'no', 'maybe']) -> None:
    match choice:
        case x if is_yes_or_no(x):
            print(f"Choice is a yes or no: {x}")
        case 'maybe':
            print("Choice is maybe.")
        case _ as never:
            assert_never(never)

check_choice('yes')
```

Here, pyright v1.1.402 on Python 3.13.5 gives the following error:

```
Argument of type "Literal['yes', 'no']" cannot be assigned to parameter "arg" of type "Never" in function "assert_never"
    Type "Literal['yes', 'no']" is not assignable to type "Never"
      Type "Literal['no']" is not assignable to type "Never" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

**Describe the solution you’d like**
No errors with the above code, same as in `mypy` (v1.16.1). Removing the 'maybe' clause with mypy correctly gives rise to an error:

```
 error: Argument 1 to "assert_never" has incompatible type "Literal['maybe']"; expected "Never"  [arg-type]
```

I have not studied PEP 742 closely enough to determine whether this is a bug in pyright or not, but it would at least be a useful feature in line with the spirit of the PEP.

**Comment by erictraut** [comment]
Thanks for the suggestion, but this doesn't strike me as a common pattern. The more typical (and less verbose) way to do this is simply:

```python
def check_choice(choice: Literal['yes', 'no', 'maybe']) -> None:
    match choice:
        case 'yes' | 'no' as x:
            print(f"Choice is a yes or no: {x}")
        case 'maybe':
            print("Choice is maybe.")
        case _ as never:
            assert_never(never)
```

Adding generalized support for type narrowing via `case` guard functions would require quite a bit of work to implement in pyright. It would be hard to justify such an investment if it's used by a handful of pyright users. I'm therefore going to reject the enhancement request and close the issue to indicate that there is currently no plan to implement this.

I'm willing to revisit this in the future if there's strong enough signal from other pyright users (in the form of upvotes on this issue).

**Comment by mdalif255** [comment]
![Image](https://github.com/user-attachments/assets/659d5ed2-ff6a-4feb-8533-3bf40281dd4f)

**Comment by sveinugu** [comment]
> Thanks for the suggestion, but this doesn't strike me as a common pattern.

Not so sure about that.

My use case for this is that I have too many options for a manual or-pattern listing to make sense in light of clean coding practices, and this is the only way I have found to move the membership check for subsets of options to an external function for reuse in slightly different pattern matching contexts.

Python devs might be pushing `Enum` for this purpose, but I am trying to move away from using `Enum` to instead use `Literal,` as IMHO the implementation of Enums in Python is somewhat broken. This is especially true if you want to merge subgroups, since it has been decided that it should not be possible to extend enums by inheritance (a decision where it seems typing concerns have been prioritised above general usefulness). Literals on the other hand have inherent support for subgroups through transitive Unions (e.g. `Literal[Literal['yes', 'no'], Literal['maybe']] == Literal['yes', 'no', 'maybe']`). Having direct static checks of value assignments is also a plus.

It seems the community to some degree favours `Literal` over `Enum`, see e.g. top comment of https://www.reddit.com/r/Python/comments/1g52103/comment/ls7xqcq/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

The lack of exhaustiveness checks for the above pattern becomes an issue when you have several large subsets of options you would like to combine, while maintaining similar type of guarantees from static typing checks as with Enums (to e.g. make sure that you have not forgotten to include some options).

I would be happy to learn about other solutions to my issue (but I also understand that this is not a forum for general discussions).

**Comment by mdalif255** [comment]
from typing import Literal, TypeIs, Any, assert_never

def is_yes_or_no(choice: Any) -> TypeIs[Literal['yes', 'no']]:
    return choice in ['yes', 'no']

def check_choice(choice: Literal['yes', 'no', 'maybe']) -> None:
    match choice:
        case x if is_yes_or_no(x):
            print(f"Choice is a yes or no: {x}")
        case 'maybe':
            print("Choice is maybe.")
        case _ as never:
            assert_never(never)

check_choice('yes')
